File: Algorithm/Java/120. Triangle.java
/*
 * Created on Sat Mar 31 2018 20:40:15
 * Author: WuLC
 * EMail: liangchaowu5@gmail.com
 */

// dp, O(n) space, from bottom to top
class Solution 
{
    public int minimumTotal(List<List<Integer>> triangle) 
    {
        int n = triangle.size();
        if (n == 0) return 0;
        int[] curr = new int[n];
        for(int i=n-1; i>=0; i--)
        {
            for(int j=0; j<triangle.get(i).size(); j++)
            {
                if(i==n-1) curr[j] = triangle.get(i).get(j);
                else curr[j] = Math.min(curr[j], curr[j+1]) + triangle.get(i).get(j);
            }
        }
        return curr[0];
    }
}

File: Algorithm/C++/120. Triangle.cpp
/*
 * Created on Tue May 08 2018 21:45:31
 * Author: WuLC
 * EMail: liangchaowu5@gmail.com
 */


// dp, O(n) space
class Solution 
{
    public:
        int minimumTotal(vector<vector<int>>& triangle) 
        {
            int m = triangle.size();
            int dp[m] = {0};
            for(int i=0; i<m; i++)
            {
                for(int j=i; j>=0; j--)
                {
                    if(j==0) dp[j] = triangle[i][j] + dp[j];
                    else if (j==i) dp[j] = triangle[i][j] + dp[j-1];
                    else dp[j] = triangle[i][j] + std::min(dp[j],dp[j-1]);
                }
            }
            int result = dp[0];
            for(int i=1; i<m; i++) result = std::min(result, dp[i]);
            return result;
        }
};